shared/core/types/relay_test.go

```diff
@@ -0,0 +1,137 @@
+package types
+
+import (
+	"errors"
+	"testing"
+)
+
+func TestRelayValidate(t *testing.T) {
+	testCases := []struct {
+		name     string
+		relay    Relay
+		expected error
+	}{
+		{
+			name: "valid Relay: JSONRPC",
+			// IMPROVE: use a factory function to build test relays
+			relay: Relay{
+				RelayPayload: &Relay_JsonRpcPayload{
+					JsonRpcPayload: &JSONRPCPayload{JsonRpc: "2.0", Method: "eth_blockNumber"},
+				},
+			},
+		},
+		{
+			name: "valid Relay: REST",
+			relay: Relay{
+				RelayPayload: &Relay_RestPayload{
+					RestPayload: &RESTPayload{Contents: `{"field1": "value1", "field2": "value2"}`},
+				},
+			},
+		},
+		{
+			name:     "invalid Relay: missing payload",
+			expected: errInvalidRelayInvalidPayload,
+		},
+		{
+			name: "invalid Relay: invalid JSONRPC Payload",
+			relay: Relay{
+				RelayPayload: &Relay_JsonRpcPayload{
+					JsonRpcPayload: &JSONRPCPayload{JsonRpc: "foo"},
+				},
+			},
+			expected: errInvalidJSONRPCInvalidRPC,
+		},
+		{
+			name: "invalid Relay: invalid REST Payload",
+			relay: Relay{
+				RelayPayload: &Relay_RestPayload{
+					RestPayload: &RESTPayload{Contents: "foo"},
+				},
+			},
+			expected: errInvalidRESTPayload,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			err := testCase.relay.Validate()
+			switch {
+			case err == nil && testCase.expected != nil:
+				t.Fatalf("Expected error %v, got: nil", testCase.expected)
```

Comment:
We use the `testify` package, which does automatic "expected vs got" type comments. https://pkg.go.dev/github.com/stretchr/testify/require#Error

You should be able to do:

```go
if testCase.expected == nil {
	require.NoError(t, err)
} else {
	require.ErrorIs(t, err, testCase.expected)
}
```

Ditto below

Reply:
Thank you for the review. Fixed.